### PR TITLE
Warn when feature / glue paths passed as file schemes

### LINKED ISSUE #2543: Warn when src/main or src/test is used in a glue or feature path
### 👓 What did you see?

Novice users often do not understand the difference between the classpath and the filesystem path.

For example this mistake often shows up on stack overflow:

```java
@RunWith(Cucumber.class)
@CucumberOptions(
        features = "src/test/resources/features",
        glue = "src/test/java/com/example"
)
```

This should be

```java
@RunWith(Cucumber.class)
@CucumberOptions(
        features = "classpath:features",
        glue = "com.example"
)
```

Here `scr/main/java/com/example` is a filesystem path while `classpath:com/example` a classpath path.

Due to misunderstanding the difference between these paths novice users often specify paths that are positively not glue paths.

Feature paths often just happen to work because unlike glue features don't have to be on the classpath.

Additionally while classpath uris are supported for glue paths the package name format is preferred to avoid confusion.

### ✅ What did you expect to see?

When passing any feature path starting with `src/{test,main}/resources` emit a warning that suggests replacing it with `classpath:` prefix.

Whe passing a glue path starts with  `src/{test,main}/{java,kotlin,scala,groovy}` emit a warning that advices the use of the package name.

The warning should include the original path used and the suggested replacement.

For the feature path it should also include an explanation that mentions the possible ambiguity and advice the user to avoid it. Either by using the suggested replacement or moving the feature files away from well known filesystem paths that are put on the classpath.

For glue paths it should inform users that files in these paths are typically compiled by their build tool before the compiled files are put on the classpath when executing. As such they probably meant to use the package name.

### 🛠️ Implementation hints

Feature path parsing is implemented in [`FeaturePath`](https://github.com/cucumber/cucumber-jvm/blob/main/core/src/main/java/io/cucumber/core/feature/FeaturePath.java). Checking for the prefix in `parseAssumeFileScheme` seems like the right place.

Glue paths are parsed in [`GluePath`](https://github.com/cucumber/cucumber-jvm/blob/main/core/src/main/java/io/cucumber/core/feature/GluePath.java). The `parseAssumeClasspathScheme` seems like the right place.

Logging warnings should be done using Cucumbers logging utility.  For example see: https://github.com/cucumber/cucumber-jvm/blob/549fe862b42ac3f0d7aab0a4cc91f05052ef737c/core/src/main/java/cucumber/api/cli/Main.java#L29

### :test_tube: Testing hints

You can test log messages by adding a `LogRecordListener` to the `LoggerFactory`. See the surrounding code for more context:

https://github.com/cucumber/cucumber-jvm/blob/78adb8fe2e4c9d80ccfc36b1a69a8a27bab4ace7/core/src/test/java/io/cucumber/core/runtime/FeaturePathFeatureSupplierTest.java#L30-L39

### PR DESCRIPTION
**Pull request by tal66**
Hi, this closes #2543. changes:
- warn when 'src/{test,main}/resources' used in features path, 
- warn when 'src/{test,main}/{java,kotlin,scala,groovy}' used in glue path
- tests for the above
- had to change core/runtime/FeaturePathFeatureSupplierTest as well, because it checked log messages at a specific index - which now changed. so also added a more convenient static method for this in LogRecordListener, used in my tests as well.

### CODE REVIEW

File: core/src/main/java/io/cucumber/core/logging/LogRecordListener.java [hunk 1]
```diff
@@ -17,4 +17,9 @@ public List<LogRecord> getLogRecords() {
         return Arrays.asList(logRecords.toArray(new LogRecord[0]));
     }
 
+    public static boolean anyRecordMessageMatch(LogRecordListener listener, String regex) {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/logging/LogRecordListener.java:20]: You can rewrite any `static` method that has an instance of its own class as an argument to an instance method.


```java
    public boolean anyRecordMessageMatch(String regex) { }
```

And then invoke as

```
logRecordListener.anyRecordMessageMatch(".*");
```

File: core/src/main/java/io/cucumber/core/feature/GluePath.java [hunk 2]
```diff
@@ -29,6 +34,11 @@
  */
 public class GluePath {
 
+    private static final Logger log = LoggerFactory.getLogger(GluePath.class);
+
+    private static final Pattern FILESYSTEM_PATH_TO_PACKAGES = Pattern
+            .compile("/*src/(?:main|test)/(?:java|kotlin|scala|groovy)/(.*)");
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/feature/GluePath.java:40]: This pattern suggests that `/src/main/java/com/example` is a file system path to a package. But I think it is unlikely that the file system root is part of any project. So I don't think we have to warn users about this. They'll be warned did not find any glue.

The pattern also suggests that `src/main/java` is not a path to a package (it's the root package). Consider adding a test case for this.

File: core/src/main/java/io/cucumber/core/feature/FeaturePath.java [hunk 3]
```diff
@@ -27,6 +32,10 @@
  */
 public class FeaturePath {
 
+    private static final Logger log = LoggerFactory.getLogger(FeaturePath.class);
+
+    private static final Pattern FILESYSTEM_PATH_TO_RESOURCES = Pattern.compile("/*src/(?:main|test)/resources/(.*)");
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/feature/FeaturePath.java:37]: This pattern suggests that `/src/test/resources/com/example` is a file system path to a package. But I think it is unlikely that the file system root is part of any project. So I don't think we have to warn users about this. They'll be warned about cucumber not being able to find any feature files.

The pattern also suggests that `src/test/resources` is not a path to a package (it's the root package). Consider adding a test case for this.

File: core/src/main/java/io/cucumber/core/feature/FeaturePath.java [hunk 4]
```diff
@@ -101,4 +111,13 @@ private static String normalize(final String value) {
         return value.toLowerCase(Locale.US).replaceAll("[^a-z0-9]+", "");
     }
 
+    private static void warnWhenFileSystemPathToResources(String featureIdentifier) {
+        Matcher matcher = FILESYSTEM_PATH_TO_RESOURCES.matcher(featureIdentifier);
+        if (matcher.matches()) {
+            log.warn(() -> String.format("Please replace feature path '%s' with the classpath '%s' to avoid ambiguity.",
+                featureIdentifier,
+                matcher.replaceAll("classpath:$1")));
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/feature/FeaturePath.java:119]: I think we're missing a bit of explanation here. Some one reading this might wonder:

* What did I do wrong?
* Why should I follow this suggestion, my code works right?

File: core/src/test/java/io/cucumber/core/options/CucumberOptionsAnnotationParserTest.java [hunk 5]
```diff
@@ -218,6 +220,44 @@ void create_with_glue() {
...
+        GluePath.parse("src/main/java/com/example");
+        GluePath.parse("src/test/java/com/package/other_package");
+
+        LoggerFactory.removeListener(logRecordListener);
+
+        assertAll(
+            () -> assertTrue(LogRecordListener.anyRecordMessageMatch(logRecordListener,
+                ".*replace.*src/main/java/com/example.*'com.example'.*")),
+            () -> assertTrue(LogRecordListener.anyRecordMessageMatch(logRecordListener,
+                ".*replace.*src/test/java/com/package/other_package.*'com.package.other_package'.*")));
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/options/CucumberOptionsAnnotationParserTest.java:258]: When this assertion fails, the assertion failed exception will not contain any useful information. Consider using matchers instead. They usually generate better exception messages. For example:

```
var records = logRecordListener.getLogRecords().stream()
  .map(LogRecord::getMessage)
  .collect(toList());
  
asserThat(records, containsInAnyOrder(matchesPattern(".*some-pattern.*)));
```  

See:

 * http://hamcrest.org/JavaHamcrest/javadoc/2.2/org/hamcrest/collection/IsIterableContainingInAnyOrder.html#containsInAnyOrder-org.hamcrest.Matcher...-
 * http://hamcrest.org/JavaHamcrest/javadoc/2.2/org/hamcrest/Matchers.html#matchesPattern-java.util.regex.Pattern-

File: core/src/test/java/io/cucumber/core/options/CucumberOptionsAnnotationParserTest.java [hunk 6]
```diff
@@ -218,6 +220,44 @@ void create_with_glue() {
...
+
+    @Test
+    void warn_when_glue_as_filesystem_path() {
+        // warn when 'src/{test,main}/{java,kotlin,scala,groovy}' is used
+
+        LogRecordListener logRecordListener = new LogRecordListener();
+        LoggerFactory.addListener(logRecordListener);
+
+        GluePath.parse("src/main/java/com/example");
+        GluePath.parse("src/test/java/com/package/other_package");
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/options/CucumberOptionsAnnotationParserTest.java:250]: Because we're testing multiple inputs for a function of `path -> log message` consider using [Parameterized tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests), specifically the `testWithMultiArgMethodSource` example.

This makes it possible to test multiple different inputs/outputs for a function while also having a single test for each case.

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
Thanks! That was quick!

Looks like we're going in the right direction over all but there are a few things to consider:

* It is usually a good idea to keep the tests as close to the code under test as possible. There are dedicated unit tests in `FeaturePathTest` and `GluePathTest`. Consider moving the tests there.

* I do think the `anyRecordMatch` method is pretty neat but I also think that after adding parameterized tests and fixing the root cause of the failure in the `FeaturePathFeatureSupplierTest` it won't be needed anymore. If done right there should only be one log record per test case.

File: core/src/test/java/io/cucumber/core/options/CucumberOptionsAnnotationParserTest.java (hunk 5 above)
Comment (mpkorstanje) [reply core/src/test/java/io/cucumber/core/options/CucumberOptionsAnnotationParserTest.java:258]: And unfortunately we're still using `Hamcrest`. If you're familiar with `AssertJ` feel free to use that.

**Comment by tal66** [comment]
thanks for the feedback, changed accordingly (except for adding explanation about the feature path, because it does work as file scheme, so not sure what more to say)

**Comment by mpkorstanje** [comment]
No problem. I'll think about the explanation a bit.

**Review by mpkorstanje — COMMENTED** [review]
I think we're almost there now. Currently the regex patterns also match a few things they shouldn't match. Consider adding tests where nothing is logged. 

Hint: You can pass matchers as an argument to the test method.

**Comment by tal66** [comment]
changed it. 
but maybe reconsider whether the feature path warning is needed at all? if it's just a best practice, and has no noticeable effect on the execution, than maybe a warning is not the place to mention that. especially if someone already put their feature files under resources (and those with `src/main/resources-hello` don't get a warning).

File: core/src/main/java/io/cucumber/core/feature/GluePath.java [hunk 7]
```diff
@@ -29,6 +34,11 @@
  */
 public class GluePath {
 
+    private static final Logger log = LoggerFactory.getLogger(GluePath.class);
+
+    private static final Pattern FILESYSTEM_PATH_TO_PACKAGES = Pattern
+            .compile("src/(?:main|test)/(?:java|kotlin|scala|groovy)(|/|/.+)");
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/feature/GluePath.java:40]: By having the capture group match either `''`, `'/'` or '/anything else'` we can avoid checking a few edge cases in the warning. We now know the string is either empty or starts with a `/`.

File: core/src/test/java/io/cucumber/core/feature/GluePathTest.java [hunk 8]
```diff
@@ -111,7 +127,44 @@ void absolute_windows_path_form_is_not_valid() {
         Executable testMethod = () -> GluePath.parse("C:\\com\\example\\app");
         IllegalArgumentException actualThrown = assertThrows(IllegalArgumentException.class, testMethod);
         assertThat("Unexpected exception message", actualThrown.getMessage(), is(equalTo(
-            "The glue path must have a classpath scheme C:/com/example/app")));
+                "The glue path must have a classpath scheme C:/com/example/app")));
+    }
+
+    @ParameterizedTest
+    @MethodSource("GluePathAndPatternProvider")
+    void warn_when_glue_as_filesystem_path(String gluePath, Matcher<String> logPattern) {
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/feature/GluePathTest.java:135]: Passing a matcher here instead of a string gives us some flexibility. We can now pass any sort of check. This means that we don't need to have an empty string if there is no message.

**Review by mpkorstanje — COMMENTED** [review]
I've pushed a few improvements to the `GluePath`. Please let me know what you think about the explanation.

I still have to look at the `FeaturePath`. Hopefully that will come later.

**Comment by tal66** [comment]
> let me know what you think about the explanation

I prefer shorter explanations, that start with a short action "change `<x>` to `<y>`" and might continue with a longer explanations for those interested. but it's a style thing, and your choice :)

**Comment by mpkorstanje** [comment]
That's a good idea.
 - Put the action first in the message. Then the explanation.
 - Removed the warning for `FeaturePath`. While it's not quite right, to reference files in `src/{main,test}/resources` because the build tool might still process them, it is a bit pedantic.
 
 If you can add an entry in the changelog under "Added"  I think we're done.

**Comment by tal66** [comment]
closing this?

**Comment by mpkorstanje** [comment]
Will be soon. It's good!

**Review by mpkorstanje — APPROVED** [review]

**Comment by aslakhellesoy** [comment]
Hi @tal66,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by mpkorstanje** [comment]
Cheer @tal66! Sorry for the delay. Very happy with this. If you're looking for more to do, you may like #2546.
